File: sdks/python/src/echo_agent_sdk/intervention_values.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class InterventionResult:
    """Immutable intervention decision; callback execution remains Host-owned."""

    block: bool = False
    block_reason: str | None = None
    injected_context: str | None = None
    redirect_to: str | None = None
    cancel: bool = False
    modified_args: Any = None

    @classmethod
    def allow(cls) -> InterventionResult:
        return cls()

    @classmethod
    def block_with_reason(cls, reason: str) -> InterventionResult:
        if not reason.strip():
            raise ValueError("block reason must not be empty")
        return cls(block=True, block_reason=reason)

    @classmethod
    def inject(cls, context: str) -> InterventionResult:
        return cls(injected_context=context)

    @classmethod
    def cancelled(cls) -> InterventionResult:
        return cls(cancel=True)

    @classmethod
    def modify_args(cls, args: Any) -> InterventionResult:
        return cls(modified_args=args)

File: sdks/typescript/src/intervention_values.ts
import type { ParamValue } from "./helpers.js";

/** Immutable intervention decision values; execution remains Host-owned. */
export type InterventionResult = Readonly<{
  block: boolean;
  blockReason?: string;
  injectedContext?: string;
  redirectTo?: string;
  cancel: boolean;
  modifiedArgs?: ParamValue;
}>;

export const InterventionResult = Object.freeze({
  allow(): InterventionResult {
    return Object.freeze({ block: false, cancel: false });
  },
  block(reason: string): InterventionResult {
    if (reason.trim().length === 0) throw new TypeError("block reason must not be empty");
    return Object.freeze({ block: true, blockReason: reason, cancel: false });
  },
  inject(context: string): InterventionResult {
    return Object.freeze({ block: false, injectedContext: context, cancel: false });
  },
  cancel(): InterventionResult {
    return Object.freeze({ block: false, cancel: true });
  },
  modifyArgs(args: ParamValue): InterventionResult {
    return Object.freeze({ block: false, cancel: false, modifiedArgs: args });
  },
});
